File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t or len(s) < len(t):
            return ""

        # Using a flat 128-slot list for ASCII indexing is much faster
        # than a standard dictionary in competitive sliding windows.
        target_counts = [0] * 128
        window_counts = [0] * 128

        required_matches = 0
        for char in t:
            idx = ord(char)
            if target_counts[idx] == 0:
                required_matches += 1
            target_counts[idx] += 1

        left = 0
        right = 0
        formed_matches = 0

        min_len = float('inf')
        min_left_start = 0

        s_len = len(s)

        while right < s_len:
            right_char_idx = ord(s[right])
            window_counts[right_char_idx] += 1

            if target_counts[right_char_idx] > 0 and window_counts[right_char_idx] == target_counts[right_char_idx]:
                formed_matches += 1

            while formed_matches == required_matches:
                current_len = right - left + 1
                if current_len < min_len:
                    min_len = current_len
                    min_left_start = left

                left_char_idx = ord(s[left])
                window_counts[left_char_idx] -= 1

                if target_counts[left_char_idx] > 0 and window_counts[left_char_idx] < target_counts[left_char_idx]:
                    formed_matches -= 1
                left += 1
            
            right += 1

        return "" if min_len == float('inf') else s[min_left_start : min_left_start + min_len]

File: 0076-minimum-window-substring/0076-minimum-window-substring.c
#include <string.h>
#include <stdlib.h>
#include <limits.h>

char* minWindow(char* s, char* t) {
    int sLen = strlen(s);
    int tLen = strlen(t);
    if (sLen < tLen || tLen == 0) return "";

    int targetCounts[128] = {0};
    int windowCounts[128] = {0};

    int requiredMatches = 0;
    for (int i = 0; i < tLen; i++) {
        unsigned char ch = t[i];
        if (targetCounts[ch] == 0) requiredMatches++;
        targetCounts[ch]++;
    }

    int left = 0, right = 0, formedMatches = 0;
    int minLen = INT_MAX;
    int minLeftStart = 0;

    while (right < sLen) {
        unsigned char rightChar = s[right];
        windowCounts[rightChar]++;

        if (targetCounts[rightChar] > 0 && windowCounts[rightChar] == targetCounts[rightChar]) {
            formedMatches++;
        }

        while (formedMatches == requiredMatches) {
            int currentLen = right - left + 1;
            if (currentLen < minLen) {
                minLen = currentLen;
                minLeftStart = left;
            }

            unsigned char leftChar = s[left];
            windowCounts[leftChar]--;

            if (targetCounts[leftChar] > 0 && windowCounts[leftChar] < targetCounts[leftChar]) {
                formedMatches--;
            }
            left++;
        }
        right++;
    }

    if (minLen == INT_MAX) return "";

    // Allocate exact memory block for output string substring (+1 for null terminator)
    char* result = (char*)malloc((minLen + 1) * sizeof(char));
    strncpy(result, s + minLeftStart, minLen);
    result[minLen] = '\0';

    return result;
}
